### PR TITLE
Move `Code.string_to_quoted` error logic up

### PR DESCRIPTION
**Pull request by drincruz**
As discussed in #7270, it would help if we moved up the error logic out of Erlang. So, instead of raising an exception, we'll return an error tuple.
Currently, this will only fix the Hex error use case, so does not completely resolve this issue. But, if this is the route we want to take, I can work on the other usecases.

### CODE REVIEW

File: lib/elixir/lib/code.ex [hunk 1]
```diff
@@ -661,8 +661,12 @@ defmodule Code do
     file = Keyword.get(opts, :file, "nofile")
     line = Keyword.get(opts, :line, 1)
 
-    with {:ok, tokens} <- :elixir.string_to_tokens(to_charlist(string), line, file, opts) do
-      :elixir.tokens_to_quoted(tokens, file, opts)
+    case :elixir.string_to_tokens(to_charlist(string), line, file, opts) do
```
Comment (josevalim) [inline lib/elixir/lib/code.ex]: Just a note: the code before and after are equivalent. However, your version is clearer, so let's stick with it. :)

File: lib/elixir/src/elixir_interpolation.erl [hunk 2]
```diff
@@ -78,9 +78,24 @@ unescape_tokens(Tokens) ->
   unescape_tokens(Tokens, fun unescape_map/1).
 
 unescape_tokens(Tokens, Map) ->
-  [unescape_token(Token, Map) || Token <- Tokens].
```
Comment (josevalim) [inline lib/elixir/src/elixir_interpolation.erl]: This is one of the few cases where I think a `throw` would be clearer. So something like this:

```erlang
unescape_tokens(Tokens, Map) ->
  try [unescape_token(Token, Map) || Token <- Tokens] of
    Unescaped -> {ok, Unescaped}
  catch
    {error, _Msg} = Error -> Error
  end.
```

This way you can keep the implementation the same, all you need to do is to replace the ArgumentError by a throw.

File: lib/elixir/src/elixir_tokenizer.erl [hunk 3]
```diff
@@ -609,8 +609,13 @@ handle_strings(T, Line, Column, H, Scope, Tokens) ->
       Token = {Key, {Line, Column - 1, nil}, Unescaped},
       tokenize(Rest, NewLine, NewColumn + 1, Scope, [Token | Tokens]);
     {NewLine, NewColumn, Parts, Rest} ->
-      Token = {string_type(H), {Line, Column - 1, nil}, unescape_tokens(Parts, Scope)},
-      tokenize(Rest, NewLine, NewColumn, Scope, [Token | Tokens])
+      UnescapedToken = unescape_tokens(Parts, Scope),
+      case UnescapedToken of
+        {error, _Msg} = Error -> Error;
```
Comment (josevalim) [inline lib/elixir/src/elixir_tokenizer.erl]: I don't think we should introduce a new error tuple which is {error, String}. Let's change this line so it returns: `{error, {Line, Column, Message, string_type(H)}, Rest, Tokens}`. This way you won't have to change any of the calls to `:elixir.string_to_token` and ultimately the code will be simpler.

Comment (fertapric) [inline lib/elixir/src/elixir_tokenizer.erl]: Checking the coding style of this file, it seems right-hand clauses are only kept in the same line if all the clauses are kept as single lines. The same applies to the `unescaped_tokens` function in `lib/elixir/src/elixir_interpolation.erl`

File: lib/elixir/test/elixir/code_test.exs [hunk 4]
```diff
@@ -118,6 +118,11 @@ defmodule CodeTest do
       assert string_to_quoted.("foo + bar") == {:ok, {:+, [line: 1, column: 5], [foo, bar]}}
     end
 
+    test "returns an error tuple on hex errors" do
+      assert Code.string_to_quoted("\"\\x\"") ==
+               {:error, "missing hex sequence after \\x, expected \\xHH"}
```
Comment (josevalim) [inline lib/elixir/test/elixir/code_test.exs]: This is changing the contract of `Code.string_to_quoted` which returns errors in the format `{error, Tuple}`. So let's proceed as [described here](https://github.com/elixir-lang/elixir/pull/7809/files#r198758547) so we can keep all the error formats the same!

File: lib/elixir/src/elixir_tokenizer.erl (hunk 3 above)
Comment (drincruz) [reply lib/elixir/src/elixir_tokenizer.erl]: thanks! I'm a bit unfamiliar with the style guide, so I trust your judgement! I'll make updates!

File: lib/elixir/src/elixir_interpolation.erl [hunk 5]
```diff
@@ -151,8 +165,7 @@ unescape_hex(<<${, A, B, C, D, E, F, $}, Rest/binary>>, Map, Acc) when ?is_hex(A
   append_codepoint(Rest, Map, [A, B, C, D, E, F], Acc, 16);
 
 unescape_hex(<<_/binary>>, _Map, _Acc) ->
-  Msg = <<"missing hex sequence after \\x, expected \\xHH">>,
-  error('Elixir.ArgumentError':exception([{message, Msg}])).
+  throw(missing_hex_sequence).
```
Comment (drincruz) [inline lib/elixir/src/elixir_interpolation.erl]: I was not sure what we would want to throw here and am not sure how we all feel about just throwing `missing_hex_sequence`. ❔

Comment (josevalim) [reply lib/elixir/src/elixir_interpolation.erl]: This is totally fine but you can also do `throw({error, Msg})` and you catch and return the tuple later on.

Comment (drincruz) [reply lib/elixir/src/elixir_interpolation.erl]: Brilliant! That sounds like the route we'd want. Let me update! 👍

File: lib/elixir/src/elixir_interpolation.erl [hunk 6]
```diff
@@ -78,9 +78,23 @@ unescape_tokens(Tokens) ->
   unescape_tokens(Tokens, fun unescape_map/1).
 
 unescape_tokens(Tokens, Map) ->
-  [unescape_token(Token, Map) || Token <- Tokens].
+  try [unescape_token(Token, Map) || Token <- Tokens] of
```
Comment (josevalim) [inline lib/elixir/src/elixir_interpolation.erl]: We should need a try/catch  only on `unescape_tokens` as `unescape_token` itself is private.

File: lib/elixir/src/elixir.erl [hunk 7]
```diff
@@ -348,6 +348,7 @@ parser_line(Meta) ->
   end.
 
 to_binary(List) when is_list(List) -> elixir_utils:characters_to_binary(List);
+to_binary(Binary) when is_binary(Binary) -> Binary;
```
Comment (josevalim) [inline lib/elixir/src/elixir.erl]: Why is this change necessary?

File: lib/elixir/src/elixir_parser.yrl [hunk 8]
```diff
@@ -903,6 +903,8 @@ string_parts(Parts) ->
   [string_part(Part) || Part <- Parts].
 string_part(Binary) when is_binary(Binary) ->
   Binary;
+string_part({error, ErrorMsg}) ->
+  {error, ErrorMsg};
```
Comment (josevalim) [inline lib/elixir/src/elixir_parser.yrl]: This should no longer be necessary either.

File: lib/elixir/src/elixir_tokenizer.erl [hunk 9]
```diff
@@ -609,8 +609,13 @@ handle_strings(T, Line, Column, H, Scope, Tokens) ->
       Token = {Key, {Line, Column - 1, nil}, Unescaped},
       tokenize(Rest, NewLine, NewColumn + 1, Scope, [Token | Tokens]);
     {NewLine, NewColumn, Parts, Rest} ->
-      Token = {string_type(H), {Line, Column - 1, nil}, unescape_tokens(Parts, Scope)},
-      tokenize(Rest, NewLine, NewColumn, Scope, [Token | Tokens])
+      Unescaped = unescape_tokens(Parts, Scope),
```
Comment (josevalim) [inline lib/elixir/src/elixir_tokenizer.erl]: Let's return `{ok, Unescaped} | {error, Msg}`

File: lib/elixir/src/elixir_tokenizer.erl [hunk 10]
```diff
@@ -609,8 +609,13 @@ handle_strings(T, Line, Column, H, Scope, Tokens) ->
...
+      case Unescaped of
+        {error, Msg} -> {error, {Line, Column, Msg, string_type(H)}, Rest, Tokens};
```
Comment (josevalim) [inline lib/elixir/src/elixir_tokenizer.erl]: Sorry, I gave a bad advice, the fourth element should be `[H]`.

File: lib/elixir/src/elixir.erl (hunk 7 above)
Comment (drincruz) [reply lib/elixir/src/elixir.erl]: oh I was getting an error since we are passing a binary from the error message [here](https://github.com/elixir-lang/elixir/pull/7809/files#diff-8e79b6aba4faa9923e1ed2defe97e290R614).

File: lib/elixir/src/elixir_interpolation.erl (hunk 6 above)
Comment (drincruz) [reply lib/elixir/src/elixir_interpolation.erl]: @josevalim Oh! thank you for bearing with me on all of these erlang changes! 😬 i'll update.

File: lib/elixir/src/elixir.erl (hunk 7 above)
Comment (josevalim) [reply lib/elixir/src/elixir.erl]: I see. It is the same "principle" as before then. We should not introduce new "formats" for returning errors. We already have a well defined tuple, we should stick to it. :)

File: lib/elixir/src/elixir_interpolation.erl [hunk 11]
```diff
@@ -78,7 +78,13 @@ unescape_tokens(Tokens) ->
...
+    Unescaped ->
+      {ok, Unescaped}
```
Comment (drincruz) [inline lib/elixir/src/elixir_interpolation.erl]: so, with this change to return an `ok | error` tuple, i needed to update Kernel and Macro slightly to respect this change as well.

Comment (josevalim) [reply lib/elixir/src/elixir_interpolation.erl]: Yes, exactly. You should also change the macro ones to raise on error, as otherwise it would be a breaking change. So it would be something like:

```elixir
case :elixir_interpolation.unescape_tokens(...) do
  {:ok, value} -> value
  {:error, reason} -> raise ArgumentError, to_string(reason)
end
```

Comment (drincruz) [reply lib/elixir/src/elixir_interpolation.erl]: that makes sense. this is a small change, i can get this updated PR in soon. 👍

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:
